ran-core/src/test/java/io/ran/ClazzMethodTest.java

```diff
@@ -0,0 +1,535 @@
+		ClazzMethod wildcard = Clazz.of(GenericTesterImpl.class).methods().find("wildcard", List.class, String.class).get();
+		assertMethod(g(GenericTester.class, String.class), wildcard, g(List.class, Object.class), String.class);
+		assertFalse(wildcard.hasGenericFromClass());
+		assertTrue(wildcard.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testHasNeitherGenericFromMethodNorClass() {
+		// String method3(String input)
+		ClazzMethod method3 = Clazz.of(GenericTesterImpl.class).methods().find("method3", String.class, String.class).get();
+		assertMethod(g(GenericTester.class, String.class), method3, String.class, String.class);
+		assertFalse(method3.hasGenericFromClass());
+		assertFalse(method3.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testHasGenericFromClass_parameterized() {
+		// List<T> param1(List<T> input)
+		ClazzMethod param1 = Clazz.of(GenericTesterImpl.class).methods().find("param1", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), param1, g(List.class, String.class), g(List.class, String.class));
+		assertTrue(param1.hasGenericFromClass());
+		assertFalse(param1.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testHasGenericFromMethod_parameterized() {
+		// <T2> List<T2> param2(List<T2> input)
+		ClazzMethod param2 = Clazz.of(GenericTesterImpl.class).methods().find("param2", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), param2, g(List.class, Object.class), g(List.class, Object.class));
+		assertFalse(param2.hasGenericFromClass());
+		assertTrue(param2.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testParamNonGeneric() {
+		// List<String> paramNonGeneric(List<String> input)
+		ClazzMethod paramNonGeneric = Clazz.of(GenericTesterImpl.class).methods().find("paramNonGeneric", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), paramNonGeneric, g(List.class, String.class), g(List.class, String.class));
+		assertFalse(paramNonGeneric.hasGenericFromClass());
+		assertFalse(paramNonGeneric.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testHasGenericFromBothMethodAndClass_parameterized() {
+		// <T2> List<T> mixed2(List<T2> input)
+		ClazzMethod mixed2 = Clazz.of(GenericTesterImpl.class).methods().find("mixed2", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), mixed2, g(List.class, String.class), g(List.class, Object.class));
+		assertTrue(mixed2.hasGenericFromClass());
+		assertTrue(mixed2.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testHasBothGenericFromMethodAndClass_parameterizedWithClosure() {
+		// <T2 extends T> List<T2> mixed3(List<T2> input)
+		ClazzMethod mixed3 = Clazz.of(GenericTesterImpl.class).methods().find("mixed3", List.class, List.class).get();
+		assertMethod(g(GenericTester.class, String.class), mixed3, g(List.class, String.class), g(List.class, String.class));
+		assertFalse(mixed3.hasGenericFromClass());
```

Comment:
Should this not be true?

Reply:
that's debatable, I could try to make it so